test/box/net.box_closing_without_lost_gh-6338.test.lua

```diff
@@ -0,0 +1,111 @@
+end;
+function check_requests_result_and_truncate_space(space, count)
+    local rc = space:select()
+    space:truncate()
+    return (#rc == count)
+end;
+test_run:cmd("setopt delimiter ''");
+
+box.schema.user.grant('guest', 'read,write,execute,create,drop', 'universe',  nil, {if_not_exists = true})
+
+s = create_space()
+connection_send_async_requests(box.cfg.net_msg_max - 1, true)
+assert(check_requests_result_and_truncate_space(s, box.cfg.net_msg_max - 1))
+
+errinj = box.error.injection
+-- A special test to check that there is no hang when the connection is
+-- closed from the server side, when not all requests have been sent yet.
+-- (see send_cond:broadcast() usage from net_box.lua)
+cond = fiber.cond()
+channel = fiber.channel(1)
+test_run:cmd("setopt delimiter ';'")
+_ = fiber.create(function()
+    cond:wait()
+    errinj.set("ERRINJ_NETBOX_IO_ERROR", true)
+    errinj.set("ERRINJ_NETBOX_IO_DELAY", false)
+    channel:put(true)
+end);
+test_run:cmd("setopt delimiter ''");
+-- We are sending a very large number of requests here to ensure that
+-- they will not have time to send by the time the connection is closed.
+conn = net_box.connect(box.cfg.listen)
+space = conn.space.test
+errinj.set("ERRINJ_NETBOX_IO_DELAY", true)
+_ = space:replace({1}, {is_async = true})
+test_run:cmd("setopt delimiter ';'")
+cond:broadcast()
+conn:close()
+test_run:cmd("setopt delimiter ''");
+channel:get()
+errinj.set("ERRINJ_NETBOX_IO_ERROR", false)
+
+-- A special test to check that async request discards, if connection is
+-- closing and waits until send_buf became empty.
+space = nil
+not_ok = nil
+rc = nil
+test_run:cmd("setopt delimiter ';'")
+_ = fiber.create(function()
+    cond:wait()
+    not_ok, rc = pcall(
+        space.replace, space, {1}, {is_async = true}
+    )
+    errinj.set("ERRINJ_NETBOX_IO_DELAY", false)
+    channel:put(true)
+end);
+test_run:cmd("setopt delimiter ''");
+conn = net_box.connect(box.cfg.listen)
+space = conn.space.test
+errinj.set("ERRINJ_NETBOX_IO_DELAY", true)
+_ = space:replace({1}, {is_async = true})
+test_run:cmd("setopt delimiter ';'")
+cond:broadcast()
+conn:close()
+test_run:cmd("setopt delimiter ''");
+assert(not not_ok)
+-- new error: Connection is closing
+assert(rc)
+channel:get()
+
+-- A special test to check that sync request discards, if connection is
+-- closing and waits until send_buf became empty.
+conn = nil
+test_run:cmd("setopt delimiter ';'")
+_ = fiber.create(function()
+    cond:wait()
+    conn:close()
+    channel:put(true)
+end);
+test_run:cmd("setopt delimiter ''");
+conn = net_box.connect(box.cfg.listen)
+space = conn.space.test
+errinj.set("ERRINJ_NETBOX_IO_DELAY", true)
+_ = space:replace({1}, {is_async = true})
+cond:broadcast()
+test_run:cmd("setopt delimiter ';'")
+errinj.set("ERRINJ_NETBOX_IO_DELAY", false)
+not_ok, rc = pcall(space.replace, space, {1});
+test_run:cmd("setopt delimiter ''");
+assert(not not_ok)
+-- new error: Connection is closing
+assert(rc)
+channel:get()
```

Comment:
The test still seems too bulky and complicated. The test below is half its size and still tests what we need. Let's use it instead?

```lua
fiber = require('fiber')
net = require('net.box')

errinj = box.error.injection

box.schema.user.grant('guest', 'execute', 'universe')

--
-- Check that close() sends all pending requests:
--  - Block the net.box worker (with error injection).
--  - Send a few requests.
--  - Start closing the connection (in a fiber, because it blocks).
--  - Try to send another request - should fail now.
--  - Unblock the net.box worker.
--  - Check that all requests have been executed.
--
counter = 0
ch = fiber.channel()
conn = net.connect(box.cfg.listen)
errinj.set("ERRINJ_NETBOX_IO_DELAY", true)
for i = 1, 5 do \
    conn:eval('counter = counter + 1', {}, {is_async = true}) \
end
_ = fiber.new(function() conn:close() ch:put(true) end)
conn:eval('return', {}) -- fails: connection is closing
errinj.set("ERRINJ_NETBOX_IO_DELAY", false)
ch:get() -- close() returned
counter -- 5 (because all requests have been executed)

--
-- Check that close() doesn't hang on send error.
--  - Block the net.box worker (with error injection).
--  - Send a few requests.
--  - Start closing the connection (in a fiber, because it blocks).
--  - Inject IO error.
--  - Unblock the net.box worker.
--  - Check that close() completed.
--
counter = 0
ch = fiber.channel()
conn = net.connect(box.cfg.listen)
errinj.set("ERRINJ_NETBOX_IO_DELAY", true)
for i = 1, 5 do \
    conn:eval('counter = counter + 1', {}, {is_async = true}) \
end
_ = fiber.new(function() conn:close() ch:put(true) end)
errinj.set("ERRINJ_NETBOX_IO_ERROR", true)
errinj.set("ERRINJ_NETBOX_IO_DELAY", false)
ch:get()
counter -- 0 (requests were dropped because of IO error)
errinj.set("ERRINJ_NETBOX_IO_ERROR", false)

box.schema.user.revoke('guest', 'execute', 'universe')
```

Reply:
Good test, simpler then my!